File: Java/src/main/java/Singleton.java
import sun.reflect.generics.reflectiveObjects.NotImplementedException;

/**
 * Created by vladimir_afanasiev on 24.04.2017.
 */
public abstract class Singleton<T extends Object>  {
    private static volatile Object _instance;
    private static Object lock = new Object();

    protected static <T extends Object> T get_instance_internal(Class<T> clazz) throws IllegalAccessException, InstantiationException {
        if(_instance == null) {
            synchronized (lock) {
                if (_instance == null) {
                    _instance = clazz.newInstance();
                }
            }
        }
        return (T) _instance;
    }

    public static  <T extends Object> T get_instance() throws Exception {
        throw new NotImplementedException(); //Instancing is not defined in child
    };
}

File: DotNet/singletonMock/Singleton.cs
using System;
using System.Reflection;

namespace singletonMock
{
    public class Singleton<T> where T: class
    {
        private static volatile T _instance;
        private static object syncRoot = new object();

        protected Singleton(){}

        public static T Instance
        {
            get
            {
                if ( _instance == null )
                {
                    lock (syncRoot)
                    {
                        if (_instance == null)
                        {
                            _instance = Activator.CreateInstance<T>();
                        }
                    }
                }
                return _instance;
            }
        } 

    }
}